pkg/http/server_test.go

```diff
@@ -0,0 +1,472 @@
+				Expect(s.listener.Close()).To(Succeed())
+			}
+			Expect(s.tlsListener != nil).To(Equal(in.expectTLSListener))
+			if in.expectTLSListener {
+				Expect(s.tlsListener.Close()).To(Succeed())
+			}
+		},
+			Entry("with a valid http bind address", &newServerTableInput{
+				opts: Opts{
+					Handler:     handler,
+					BindAddress: "127.0.0.1:0",
+				},
+				expectedErr:        nil,
+				expectHTTPListener: true,
+				expectTLSListener:  false,
+			}),
+			Entry("with a valid https bind address, with no TLS config", &newServerTableInput{
+				opts: Opts{
+					Handler:           handler,
+					SecureBindAddress: "127.0.0.1:0",
+				},
+				expectedErr:        errors.New("error setting up TLS listener: no TLS config provided"),
+				expectHTTPListener: false,
+				expectTLSListener:  false,
+			}),
+			Entry("with a valid https bind address, and valid TLS config", &newServerTableInput{
+				opts: Opts{
+					Handler:           handler,
+					SecureBindAddress: "127.0.0.1:0",
+					TLS: &options.TLS{
+						Key:  &keyDataSource,
+						Cert: &certDataSource,
+					},
+				},
+				expectedErr:        nil,
+				expectHTTPListener: false,
+				expectTLSListener:  true,
+			}),
+			Entry("with a both a valid http and valid https bind address, and valid TLS config", &newServerTableInput{
+				opts: Opts{
+					Handler:           handler,
+					BindAddress:       "127.0.0.1:0",
+					SecureBindAddress: "127.0.0.1:0",
+					TLS: &options.TLS{
+						Key:  &keyDataSource,
+						Cert: &certDataSource,
+					},
+				},
+				expectedErr:        nil,
+				expectHTTPListener: true,
+				expectTLSListener:  true,
+			}),
+			Entry("with a \"-\" for the bind address", &newServerTableInput{
+				opts: Opts{
+					Handler:     handler,
+					BindAddress: "-",
+				},
+				expectedErr:        nil,
+				expectHTTPListener: false,
+				expectTLSListener:  false,
+			}),
+			Entry("with a \"-\" for the secure bind address", &newServerTableInput{
+				opts: Opts{
+					Handler:           handler,
+					SecureBindAddress: "-",
+				},
+				expectedErr:        nil,
+				expectHTTPListener: false,
+				expectTLSListener:  false,
+			}),
+			Entry("with an invalid bind address scheme", &newServerTableInput{
+				opts: Opts{
+					Handler:     handler,
+					BindAddress: "invalid://127.0.0.1:0",
+				},
+				expectedErr:        errors.New("error setting up listener: listen (invalid, 127.0.0.1:0) failed: listen invalid: unknown network invalid"),
+				expectHTTPListener: false,
+				expectTLSListener:  false,
+			}),
+			Entry("with an invalid secure bind address scheme", &newServerTableInput{
+				opts: Opts{
+					Handler:           handler,
+					SecureBindAddress: "invalid://127.0.0.1:0",
+					TLS: &options.TLS{
+						Key:  &keyDataSource,
+						Cert: &certDataSource,
+					},
+				},
+				expectedErr:        nil,
+				expectHTTPListener: false,
+				expectTLSListener:  true,
+			}),
+			Entry("with an invalid bind address port", &newServerTableInput{
```

Comment:
These test cases are very comprehensive - Nice!

The only addition I could think of adding is the case where a bind port is already taken.